subprojects/core/src/main/java/org/gradle/internal/model/LoadingCacheFactory.java

```diff
@@ -0,0 +1,200 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.model;
+
+import org.gradle.internal.Describables;
+import org.gradle.internal.DisplayName;
+import org.gradle.internal.service.scopes.Scope;
+import org.gradle.internal.service.scopes.ServiceScope;
+import org.gradle.internal.work.WorkerLimits;
+
+import javax.inject.Inject;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Function;
+
+/**
+ * A factory for creating {@link LoadingCache} instances of various types.
+ * <p>
+ * Caches created by this factory are intended for general use wherever concurrent
+ * in-memory caching is required.
+ */
+@ServiceScope(Scope.BuildSession.class)
+public class LoadingCacheFactory {
```

Comment:
Lets call this `CacheFactory` or `InMemoryCacheFactory`. While there are advantages of using a `LoadingCache` (providing the `computeIfAbsent` method upon construction), there are use cases where this doesn't work or is not ideal. 

This class will eventually want to provide factory methods for non-loading caches.

Reply:
Renamed to `InMemoryCacheFactory`